### PR TITLE
Add a glossary

### PR DESCRIPTION
**Pull request by Kanegraffiti**
Removed numbering from the terms
Did Alphabetical order for terms
Made sure each term had a label that could be referenced 
Edited two index files to include the glossary 

This is related to issue #3426 
Recreates changes from this closed [PR](https://github.com/Parsl/parsl/pull/3578) with corrections from @sophie-bui 

Please review here @WardLT

### CODE REVIEW

**Comment by WardLT** [comment]
A quick side note: no need to create a new PR for each set up changes. It's no problem that you have, but don't feel like you need to.

File: docs/userguide/glossary.rst [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+---------
+
+A job in Parsl refers to a unit of work that is submitted to an execution environment (like a cluster or cloud) for processing. It's like a task that needs to be done, such as running a script or processing data, and it can consist of one or more tasks that are executed on a compute resource.
+
+.. _launcher:
+
+**Launcher:**
+----------------
+
+A launcher is like a starter for your apps. It's the part that actually starts the apps running on the computers. It's like a coach telling the athletes when to start running.
```
Comment (WardLT) [inline docs/userguide/glossary.rst:113]: A point of clarity: the Launcher doesn't actually start the apps. It puts the workers onto each computer. So, more like the bus driver who brings the players to a stadium than the coach who tells them what to do.

**Comment by WardLT** [comment]
That change looks good, but it seems like there's a problem with duplicate labels. The run is failing because

```
Warning, treated as error:
/home/runner/work/parsl/parsl/docs/historical/changelog.rst:947:more than one target found for 'any' cross-reference 'DataFuture': could be :std:ref:`DataFuture:` or :py:class:`parsl.app.futures.DataFuture`
```

Does that error make sense to use? It sounds like we might need to pick a different name for the reference label for DataFuture in your glossary

**Comment by Kanegraffiti** [comment]
> That change looks good, but it seems like there's a problem with duplicate labels. The run is failing because
> 
> ```
> Warning, treated as error:
> /home/runner/work/parsl/parsl/docs/historical/changelog.rst:947:more than one target found for 'any' cross-reference 'DataFuture': could be :std:ref:`DataFuture:` or :py:class:`parsl.app.futures.DataFuture`
> ```
> 
> Does that error make sense to use? It sounds like we might need to pick a different name for the reference label for DataFuture in your glossary

I have changed the DataFuture label, I considered changing others to match but I am worried about introducing new errors

**Comment by benclifford** [comment]
>  I am worried about introducing new errors

if you get new errors, then you get new errors and then you fix them in another commit - that isn't something to be scared of. my PRs fail CI about 50% of the time. what matters is not having errors *when this PR is merged to the `master` branch*.

**Comment by Kanegraffiti** [comment]
> > I am worried about introducing new errors
> 
> if you get new errors, then you get new errors and then you fix them in another commit - that isn't something to be scared of. my PRs fail CI about 50% of the time. what matters is not having errors _when this PR is merged to the `master` branch_.

Thank you. I'll change the other labels now so they match

**Comment by sophie-bui** [comment]
@benclifford – do we need to update Kelechi's permissions in our repo so she can contribute to our docs? I saw your message in our hacker's channel, so I want to double-check.

**Comment by benclifford** [comment]
Here's an error that is happening in the documentation build now: (click on the red x of parsl / mail test-suite 3.8 and scroll all the way to the bottom)

```
Warning, treated as error:
/home/runner/work/parsl/parsl/docs/userguide/glossary.rst:99:'any' reference target not found: .result()
```

That sounds like something at line 99 of glossary.rst being treated as a reference, around the text `.result()` that you maybe don't want treated as a reference.

**Comment by Kanegraffiti** [comment]
> Here's an error that is happening in the documentation build now: (click on the red x of parsl / mail test-suite 3.8 and scroll all the way to the bottom)
> 
> ```
> Warning, treated as error:
> /home/runner/work/parsl/parsl/docs/userguide/glossary.rst:99:'any' reference target not found: .result()
> ```
> 
> That sounds like something at line 99 of glossary.rst being treated as a reference, around the text `.result()` that you maybe don't want treated as a reference.

The .result() method in the Future definition has been changed to double backticks, which should render it as inline code instead

**Comment by Kanegraffiti** [comment]
We did it! Thank you for helping @sophie-bui @WardLT @benclifford

File: docs/userguide/glossary.rst [hunk 2]
```diff
@@ -0,0 +1,218 @@
+Glossary of Parsl Terms
+=======================
+
+This glossary defines terms based on their usage within Parsl. By defining our terminology, we hope to create understanding across our community and reduce confusion. When asking for or providing support to fellow Parsl users, please use these terms as defined.
+
+Our glossary is organized alphabetically by the English alphabet. Feel free to contribute terms and definitions to this list that will benefit Parsl users.
```
Comment (danielskatz) [inline docs/userguide/glossary.rst:6]:
```suggestion
Our glossary is organized alphabetically in English. Feel free to contribute terms and definitions to this list that will benefit Parsl users.
```

File: docs/userguide/glossary.rst [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+
+.. _glossary:
+
+.. _appglossary:
+   
+**App:**
+----------
+
+In Parsl, an app is like a small, self-contained program that does a specific job. It's a piece of code, like a Python function or a Bash script, that you can run separately from your main program. Think of it as a mini-tool within your larger toolbox.
```
Comment (danielskatz) [inline docs/userguide/glossary.rst:15]:
```suggestion
In Parsl, an app is a small, self-contained program that does a specific job. It's a piece of code, such as a Python function or a Bash script, that can run separately from your main program. Think of it as a mini-tool within your larger toolbox.
```
```suggestion
In Parsl, an app is like a small, self-contained program that does a specific job. It's a piece of code, like a Python function or a Bash script, that you can run separately from your main program. Think of it as a mini-tool within your larger toolbox.
```

[the remark this answers is not in the document]
Comment (danielskatz) [reply docs/userguide/glossary.rst:15]: How is an App different than a Task, as defined below?

File: docs/userguide/glossary.rst [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+.. _appfutureglossary:
+
+**AppFuture:**
+-----------------
+
+An AppFuture is like a placeholder for the result of an app that's running in the background. It's like a ticket you get when you order food at a restaurant – you get the ticket right away, but you have to wait for the food to be ready. Similarly, you get an AppFuture immediately when you start an app, but you have to wait for the app to finish before you can see the results.
```
Comment (danielskatz) [inline docs/userguide/glossary.rst:22]:
```suggestion
An AppFuture is a placeholder for the result of an app that runs in the background. It's like a ticket you get when you order food at a restaurant – you get the ticket right away, but you have to wait for the food to be ready. Similarly, you get an AppFuture immediately when you start an app, but you have to wait for the app to finish before you can see the results.
```

File: docs/userguide/glossary.rst [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+---------------
+   
+A Bash app is a special kind of app in Parsl that lets you run commands from your computer's terminal (like the ones you type in the command prompt or shell). It's a way to use Parsl to automate tasks that you would normally do manually in the terminal.
+
+.. _blockglossary:
+
+**Block:**
+------------
+
+A block is like a group of computers loaned to you for doing work. Parsl will manage passing out work to each of them to do. This way, you can get your work done faster by using multiple computers at the same time.
```
Comment (danielskatz) [inline docs/userguide/glossary.rst:36]: Is "computers" the right term for the elements of a Block?

File: docs/userguide/glossary.rst [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+
+.. _checkpointingglossary:
+
+**Checkpointing:**
+---------------------
+
+Checkpointing is like saving your progress in a video game. If something goes wrong, you can restart from the last saved point instead of starting all over again. In Parsl, checkpointing lets you save the state of your work so that you can resume it later if there's an interruption.
```
Comment (danielskatz) [inline docs/userguide/glossary.rst:43]:
```suggestion
Checkpointing is like saving your progress in a video game. If something goes wrong, you can restart from the last saved point (checkpoint) instead of starting all over again. In Parsl, checkpointing lets you save the state of your work so that you can resume it later if there's an interruption.
```

File: docs/userguide/glossary.rst [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+
+.. _concurrencyglossary:
+
+**Concurrency:**
+-------------------
+
+Concurrency means doing multiple things at the same time. In Parsl, it means that your apps can run simultaneously, even if they're not on the same computer. This can make your programs run much faster.
```
Comment (danielskatz) [inline docs/userguide/glossary.rst:50]: Is "computer" the right word here?

File: docs/userguide/glossary.rst [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+
+.. _configurationglossary:
+
+**Configuration:**
+---------------------
+
+Configuration is like setting up the rules for how Parsl should work. It's like adjusting the settings on your phone – you can choose how you want things to look and behave. In Parsl, you can configure things like how many computers to use, where to store data, and how to handle errors.
```
Comment (danielskatz) [inline docs/userguide/glossary.rst:57]:
```suggestion
Configuration sets up the rules for how Parsl should work. It's like adjusting the settings on your phone – you can choose how you want things to look and behave. In Parsl, you can configure things like how many computers to use, where to store data, and how to handle errors.
```

Comment (danielskatz) [reply docs/userguide/glossary.rst:57]: Is "computers" the right word to use here?

File: docs/userguide/glossary.rst [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+
+.. _datafutureglossary:
+
+**DataFuture:**
+------------------
+
+A DataFuture is like a placeholder for a file that an app is creating. It's like a receipt for a package you're expecting – you get the receipt right away, but you have to wait for the package to arrive. Similarly, you get a DataFuture immediately when an app starts creating a file, but you have to wait for the file to be finished before you can use it.
```
Comment (danielskatz) [inline docs/userguide/glossary.rst:64]:
```suggestion
A DataFuture is a placeholder for a file that an app is creating. It's like a receipt for a package you're expecting – you get the receipt right away, but you have to wait for the package to arrive. Similarly, you get a DataFuture immediately when an app starts creating a file, but you have to wait for the file to be finished before you can use it.
```

File: docs/userguide/glossary.rst [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+------------------------------
+
+The DataFlowKernel is like the brain of Parsl. It's the part that controls how your apps run and how they share information. It's like the conductor of an orchestra, making sure that all the musicians play together in harmony.
+
+.. _elasticityglossary:
+
+**Elasticity:**
+-----------------
+
+Elasticity means being able to stretch or shrink. In Parsl, it means that you can easily add or remove blocks of computers as needed. This is useful if your workload changes – you can use more computers when you have a lot of work to do and fewer computers when you don't.
```
Comment (danielskatz) [inline docs/userguide/glossary.rst:78]: Is "computers" the right word here?

File: docs/userguide/glossary.rst [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+
+.. _executionproviderglossary:
+
+**Execution Provider:**
+--------------------------
+
+An execution provider is like a bridge between Parsl and the computers you want to use. It's the part that knows how to talk to different types of computers, like your laptop, a cluster, or a cloud service.
```
Comment (danielskatz) [inline docs/userguide/glossary.rst:85]: Is "computers" the right word here?

File: docs/userguide/glossary.rst [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+
+.. _executorglossary:
+
+**Executor:**
+----------------
+
+An executor is like a manager for your apps. It's the part that decides which app should run on which computer and when. It's like a traffic controller, directing the flow of apps to make sure they all get where they need to go.
```
Comment (danielskatz) [inline docs/userguide/glossary.rst:92]: Is "computer" the right word here?

File: docs/userguide/glossary.rst [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+-------------
+
+A future is a placeholder for the result of a task that hasn't finished yet. Both AppFuture and DataFuture are types of Futures. You can use the ``.result()`` method to get the actual result when it's ready.
+
+.. _jobglossary:
+
+**Job:**
+---------
+
+A job in Parsl refers to a unit of work that is submitted to an execution environment (like a cluster or cloud) for processing. It's like a task that needs to be done, such as running a script or processing data, and it can consist of one or more tasks that are executed on a compute resource.
```
Comment (danielskatz) [inline docs/userguide/glossary.rst:106]: Task is used two different ways here.

File: docs/userguide/glossary.rst [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+--------------
+
+A manager in Parsl is responsible for overseeing the execution of tasks on specific compute resources. It's like a supervisor who ensures that all workers (or workers within a block) are carrying out their tasks correctly and efficiently.
+
+.. _memoizationglossary:
+
+**Memoization:**
+-------------------
+
+Memoization is like remembering something so you don't have to do it again. In Parsl, it means that if you run an app with the same inputs multiple times, Parsl will remember the result from the first time and give it to you again instead of running the app again. This can save a lot of time.
```
Comment (danielskatz) [inline docs/userguide/glossary.rst:127]:
```suggestion
Memoization is like remembering something so you don't have to do it again. In Parsl, if you are using memoization and you run an app with the same inputs multiple times, Parsl will remember the result from the first time and give it to you again instead of running the app again. This can save a lot of time.
```

File: docs/userguide/glossary.rst [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+
+.. _mpiappglossary:    
+
+**MPI App:**
+---------------
+
+An MPI app is a special kind of app that uses a technology called Message Passing Interface (MPI) to communicate between different computers. It's like a walkie-talkie that lets different apps talk to each other.
```
Comment (danielskatz) [inline docs/userguide/glossary.rst:134]: Is "computers" the right word here?

Comment (danielskatz) [reply docs/userguide/glossary.rst:134]: The second sentence implies communication between apps, while MPI is really communication within an app.

File: docs/userguide/glossary.rst [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+
+.. _nodeglossary:
+
+**Node:**
+------------
+
+A node in Parsl is like a workstation in a factory. It's the physical or virtual machine where work gets done. Each node provides the computational power needed to run tasks, and it can host several workers who carry out the tasks.
```
Comment (danielskatz) [inline docs/userguide/glossary.rst:141]: How is a Node related to a "computer" as mentioned elsewhere in this glossary?

File: docs/userguide/glossary.rst [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+-------------------
+
+Parallelism means doing multiple things at the same time. In Parsl, it means that your apps can run simultaneously on different computers. This can make your programs run much faster.
+
+.. _parslscriptglossary:    
+
+**Parsl Script:**
+---------------------
+
+A Parsl script is a file that contains the instructions for how to run your apps in parallel. It's like a recipe that tells you what ingredients to use and how to combine them.
```
Comment (danielskatz) [inline docs/userguide/glossary.rst:155]: You might say that a Parsl script is a Python program that uses the Python library.

File: docs/userguide/glossary.rst [hunk 18]
```diff
@@ -0,0 +1,218 @@
...
+
+.. _pluginglossary:
+
+**Plugin:**
+---------------
+
+A plugin is like an add-on for Parsl. It's a piece of code that you can add to Parsl to give it new features or change how it works. It's like an extra tool that you can add to your toolbox.
```
Comment (danielskatz) [inline docs/userguide/glossary.rst:162]:
```suggestion
A plugin is an add-on for Parsl. It's a piece of code that you can add to Parsl to give it new features or change how it works. It's like an extra tool that you can add to your toolbox.
```

File: docs/userguide/glossary.rst [hunk 19]
```diff
@@ -0,0 +1,218 @@
...
+
+.. _pythonappglossary: 
+
+**Python App:**
+------------------
+
+A Python app is a special kind of app in Parsl that's written in the Python programming language. It's a way to use Parsl to run your Python code in parallel.
```
Comment (danielskatz) [inline docs/userguide/glossary.rst:169]:
```suggestion
A Python app is a special kind of app in Parsl that's written as a Python function. It's a way to use Parsl to run your Python code in parallel.
```

File: docs/userguide/glossary.rst [hunk 20]
```diff
@@ -0,0 +1,218 @@
...
+
+.. _resourceglossary:
+
+**Resource:**
+---------------
+
+A resource in Parsl refers to any computational asset that can be used to execute tasks, such as CPU cores, memory, or entire nodes. It's like the tools and materials you need to get a job done.
```
Comment (danielskatz) [inline docs/userguide/glossary.rst:176]: How is this related to "computers" as discussed elsewhere in this glossary?

File: docs/userguide/glossary.rst [hunk 21]
```diff
@@ -0,0 +1,218 @@
...
+--------------------
+
+Serialization is like packing your belongings into a suitcase so you can take them on a trip. In Parsl, it means converting your data into a format that can be sent over a network to another computer.
+
+.. _stagingglossary:    
+
+**Staging:**
+---------------
+
+Staging is like setting the stage for a play. In Parsl, it means preparing the data that your apps need before they start running. This can involve things like copying files to the right location or converting them into the right format.
```
Comment (danielskatz) [inline docs/userguide/glossary.rst:190]: I don't think of staging as preparing the data or converting them, but just of moving them. But I'm not 100% this is correct for Parsl...

File: docs/userguide/glossary.rst [hunk 22]
```diff
@@ -0,0 +1,218 @@
...
+
+.. _taskglossary:
+
+**Task:**
+------------
+
+A task in Parsl is the smallest unit of work that can be executed. It's like a single step in a larger process, where each task is part of a broader workflow or job.
```
Comment (danielskatz) [inline docs/userguide/glossary.rst:197]: How is a task related to an App?

File: docs/userguide/glossary.rst [hunk 23]
```diff
@@ -0,0 +1,218 @@
...
+
+.. _threadglossary:    
+
+**Thread:**
+-------------
+
+A thread is like a smaller part of a program that can run independently. It's like a worker in a factory who can do their job at the same time as other workers.
```
Comment (danielskatz) [inline docs/userguide/glossary.rst:204]: maybe mention parallelism here? And talk about where threads can be used?

Comment (danielskatz) [reply docs/userguide/glossary.rst:204]: Should processes also be defined?

File: docs/userguide/glossary.rst [hunk 24]
```diff
@@ -0,0 +1,218 @@
...
+
+.. _workerglossary:
+
+**Worker:**
+-------------
+
+A worker in Parsl is like an employee in the factory who does the actual work. Workers run on nodes and are responsible for executing the tasks assigned to them. Multiple workers can work on a single node, sharing the node’s resources to get the job done efficiently.
```
Comment (danielskatz) [inline docs/userguide/glossary.rst:211]: Worker and Thread seem similar in definition here, but I don't think they should be.

File: docs/userguide/glossary.rst [hunk 25]
```diff
@@ -0,0 +1,218 @@
...
+
+.. _workflowglossary:    
+
+**Workflow:**
+----------------
+
+A workflow is like a series of steps that you follow to complete a task. In Parsl, it's a way to describe how your apps should run and how they depend on each other. It's like a flowchart that shows you the order in which things need to happen.
```
Comment (danielskatz) [inline docs/userguide/glossary.rst:218]: How is a workflow related to a Parsl script or a Python program?

**Comment by Kanegraffiti** [comment]
Thank you for your detailed review, I really appreciate your help, time and effort. I am applying your corrections right now! I hope to get more feedback from you afterwards @danielskatz

**Comment by danielskatz** [comment]
> Thank you for your detailed review, I really appreciate your help, time and effort. I am applying your corrections right now! I hope to get more feedback from you afterwards @danielskatz

Please ping me when you want me to review again.

**Comment by Kanegraffiti** [comment]
> > Thank you for your detailed review, I really appreciate your help, time and effort. I am applying your corrections right now! I hope to get more feedback from you afterwards @danielskatz
> 
> Please ping me when you want me to review again.

Please can you look at this again, I have made corrections and it's ready for your review

File: docs/userguide/glossary.rst [hunk 26]
```diff
@@ -0,0 +1,226 @@
...
+-------------------
+
+Memoization is like remembering something so you don't have to do it again. In Parsl, if you are using memoization and you run an app with the same inputs multiple times, Parsl will remember the result from the first time and give it to you again instead of running the app again. This can save a lot of time.
+
+.. _mpiappglossary:    
+
+**MPI App:**
+---------------
+
+An MPI app is a specialized app that uses the Message Passing Interface (MPI) for communication between different nodes within the app. It enables different parts of the app to communicate and work together effectively. It's like a walkie-talkie that lets different parts of an app talk to each other.
```
Comment (danielskatz) [inline docs/userguide/glossary.rst:134]: MPI doesn't have to be used across nodes, it can also be used within nodes.

File: docs/userguide/glossary.rst [hunk 27]
```diff
@@ -0,0 +1,226 @@
...
+------------
+
+A node in Parsl is like a workstation in a factory. It's a physical or virtual machine that provides the computational power needed to run tasks. Each node can host several workers that execute tasks.
+
+.. _parallelismglossary:
+
+**Parallelism:**
+-------------------
+
+Parallelism means doing multiple things at the same time but not necessarily in the same location or using the same resources. In Parsl, it involves running apps simultaneously across different nodes or computational resources, accelerating program execution.
```
Comment (danielskatz) [inline docs/userguide/glossary.rst:148]: what's the difference between parallelism and concurrency?

File: docs/userguide/glossary.rst [hunk 28]
```diff
@@ -0,0 +1,226 @@
...
+---------------
+
+A plugin is an add-on for Parsl. It's a piece of code that you can add to Parsl to give it new features or change how it works. It's like an extra tool that you can add to your toolbox.
+
+.. _processglossary:
+
+**Process:**
+---------------
+
+In Parsl, a process is like a box that holds a set of tasks that can be executed together, sharing resources and context. A process in Parsl is a way to group related tasks, manage their dependencies, and optimize their execution.
```
Comment (danielskatz) [inline docs/userguide/glossary.rst:169]: I don't know that this is a relevant definition.

File: docs/userguide/glossary.rst [hunk 29]
```diff
@@ -0,0 +1,226 @@
...
+--------------------
+
+Serialization is like packing your belongings into a suitcase so you can take them on a trip. In Parsl, it means converting your data into a format that can be sent over a network to another computer.
+
+.. _stagingglossary:    
+
+**Staging:**
+---------------
+
+Staging in Parsl refers to moving data to the appropriate location before an app starts running. This ensures that all necessary data is available where it needs to be for the app to execute properly.
```
Comment (danielskatz) [inline docs/userguide/glossary.rst:197]: staging can also be moving data from that location after the app finishes.

**Comment by danielskatz** [comment]
I've provided a few more comments.  Also, I know @benclifford looked at this in terms of technically merging it, but it would be useful for him or @yadudoc or @kylechard to also review it for content.

**Comment by Kanegraffiti** [comment]
> I've provided a few more comments. Also, I know @benclifford looked at this in terms of technically merging it, but it would be useful for him or @yadudoc or @kylechard to also review it for content.

I have implemented the latest feedback, thank you. When differentiating between Parallelism and Concurrency I used an analogy with a similar theme in both definitions and compared that in Parallelism definition. Does it look okay?

**Comment by danielskatz** [comment]
Yes, the concurrency vs parallelism does look ok to me now.

**Comment by danielskatz** [comment]
I still would like one more review of the content, however

**Review by benclifford — APPROVED** [review]
